src/Logging.AzureAppServices/src/AzureBlobLoggerContext.cs

```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Microsoft.Extensions.Logging.AzureAppServices
+{
+    public readonly struct AzureBlobLoggerContext
+    {
+        public AzureBlobLoggerContext(string appName, string identifier, DateTimeOfset timestamp);
```

Comment:
Ah, well you need to define the constructor like this:
```suggestion
        public AzureBlobLoggerContext(string appName, string identifier, DateTimeOffset timestamp)
        {
            AppName = appName;
            Identifier = identifier;
            Timestamp = timestamp;
        }
```

Reply:
Hi JunTao, the body in the constructor will now compile. But now how do we wire the new struct and the new Func with the existing API such a way that existing use case is not required to define the File format, yet the new use case can define a flat structure shown in your example. Would we still need to make changes in BlobLoggerProvider? After all, it is internal override async Task WriteMessagesAsync in BlobLoggerProvider that originally forces the folder structure to be in deeply nested folder format of $"{_appName}/{key.Year}/{key.Month:00}/{key.Day:00}/{key.Hour:00}/{_fileName}".

If we change BlobLoggerProvider to use the option context, then the existing applications would break because none of them are using the option context up to today.

Reply:
A couple of implementation notes, let me know if these answers your question:
1. Set the default value of the `FileNameFormat` property on the options class to the logic at https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L66. i.e. something like:
```c# 
public class AzureBlobLoggerOptions
{
...
   public Func<AzureBlobLoggerContext, string> FileNameFormat { get; set; } = context =>
      $"{context.AppName}/{context.Timestamp.Year}/{context.Timestamp.Month:00}/{context.Timestamp.context.Timestamp:00}/{context.Timestamp.Hour:00}/{context.Identifier}";
...
}
```
2. We store the option monitor on the class as a private member: https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L53
3. The logic here: https://github.com/dotnet/aspnetcore/blob/9819c08be913f918daf130e85ac6fd8e3aad46c9/src/Logging.AzureAppServices/src/BlobLoggerProvider.cs#L66 will be modified to something like:
```c#
var blobName = _options.CurrentValue.FileNameFormat(new AzureBlobLoggerContext(_appName, _fileName, new DateTimeOffset (key.Year, key.Month, key.Day, key.Hour, 0, 0, TimeSpan.Zero)));
```

This way, the current behaviour is preserved and users can use the options to override the format of the file produced to anything they want, including a flat structure. See my usage example in https://github.com/dotnet/aspnetcore/pull/27383#issuecomment-822799282. Note, an overload of `AddAzureWebAppDiagnostics` that takes the action delegate needs to be added.